crates/services/gas_price_service/src/v0/uninitialized_task.rs

```diff
@@ -0,0 +1,277 @@
+        ))?;
+
+    let mut algo_updater = metadata.into();
+
+    sync_v0_metadata(
+        settings,
+        on_chain_db,
+        metadata_height,
+        latest_block_height,
+        &mut algo_updater,
+        metadata_storage,
+    )?;
+
+    Ok(())
+}
+
+fn sync_v0_metadata<L2DataStore, L2DataStoreView, GasPriceStore, SettingsProvider>(
+    settings: &SettingsProvider,
+    on_chain_db: &L2DataStoreView,
+    metadata_height: u32,
+    latest_block_height: u32,
+    updater: &mut AlgorithmUpdaterV0,
+    metadata_storage: &mut StructuredStorage<GasPriceStore>,
+) -> anyhow::Result<()>
+where
+    L2DataStore: L2Data,
+    L2DataStoreView: AtomicView<LatestView = L2DataStore>,
+    GasPriceStore: GasPriceData + Modifiable + KeyValueInspect<Column = GasPriceColumn>,
+    SettingsProvider: GasPriceSettingsProvider,
+{
+    let first = metadata_height.saturating_add(1);
+    let view = on_chain_db.latest_view()?;
+    for height in first..=latest_block_height {
+        let block = view
+            .get_block(&height.into())?
+            .ok_or(not_found!("FullBlock"))?;
+        let param_version = block.header().consensus_parameters_version;
+
+        let GasPriceSettings {
+            gas_price_factor,
+            block_gas_limit,
+        } = settings.settings(&param_version)?;
+        let block_gas_capacity = block_gas_limit.try_into()?;
+
+        let block_gas_used =
+            match get_block_info(&block, gas_price_factor, block_gas_limit)? {
+                BlockInfo::GenesisBlock => {
+                    Err(anyhow::anyhow!("should not be genesis block"))?
+                }
+                BlockInfo::Block { gas_used, .. } => gas_used,
+            };
+
+        updater.update_l2_block_data(height, block_gas_used, block_gas_capacity)?;
+        let metadata: UpdaterMetadata = updater.clone().into();
+        metadata_storage.set_metadata(&metadata)?;
+    }
+
+    Ok(())
+}
+
+pub fn new_gas_price_service_v0<
+    L2DataStore,
+    L2DataStoreView,
+    GasPriceStore,
+    SettingsProvider,
+>(
+    config: GasPriceServiceConfig,
+    genesis_block_height: BlockHeight,
+    settings: SettingsProvider,
+    block_stream: BoxStream<SharedImportResult>,
+    gas_price_db: GasPriceStore,
+    on_chain_db: L2DataStoreView,
+) -> anyhow::Result<
+    ServiceRunner<
+        GasPriceServiceV0<
+            FuelL2BlockSource<SettingsProvider>,
+            StructuredStorage<GasPriceStore>,
+        >,
+    >,
+>
+where
+    L2DataStore: L2Data,
+    L2DataStoreView: AtomicView<LatestView = L2DataStore>,
+    GasPriceStore: GasPriceData + Modifiable + KeyValueInspect<Column = GasPriceColumn>,
+    SettingsProvider: GasPriceSettingsProvider,
+{
+    let gas_price_init = UninitializedTask::new(
+        config,
+        genesis_block_height,
+        settings,
+        block_stream,
+        gas_price_db,
+        on_chain_db,
+    )?;
+    let gas_price_service_v0 = gas_price_init.init()?;
```

Comment:
Isn't this blocking now? The purpose of the `InitializeTask` was to avoid blocking the initialization of all the sub services.

Reply:
yup, its blocking now. if we want it to be non-blocking, I can - 
a. move some things around and move all the logic in the `init` fn to the `into_task` fn of the `GasPriceService`

or

b. implement `RunnableService` for `UninitializedTask` like how it is done for `p2p` and the task it returns is `GasPriceService`. this will require initialization of the shared state and have it as an additional parameter to the `GasPriceService::new()` fn call.

which direction would you prefer?

Reply:
also, in favor of the blocking argument: there are other modules which depend on the shared state being initialized, and this behaviour is more deterministic :)

Reply:
I think we should go with b. to be consistent. I like the idea of being blocking, just to control the order of setup for the sub services better, but that's not what we're doing elsewhere so let's avoid it for now.